firestore/client/field_path.h

```diff
@@ -0,0 +1,73 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLECLOUDCPP_FIRESTORE_GOOGLE_FIRESTORE_FIELD_PATH_H_
+#define GOOGLECLOUDCPP_FIRESTORE_GOOGLE_FIRESTORE_FIELD_PATH_H_
+
+#include <iostream>
+#include <regex>
+#include <string>
+#include <vector>
+
+namespace firestore {
+/**
+ * A FieldPath refers to a field in a document. The path may consist of
+ * a single field name (referring to a top level field in the document),
+ * or a list of field names (referring to a nested field in the document).
+ */
+class FieldPath {
+ public:
+  // Construct FieldPath from a vector of field names
+  FieldPath(const std::vector<std::string>& parts);
```

Comment:
The rest of the project uses `std::vector<std::string> const&`.  C++ types are read from right to left: a reference to a constant vector of strings.